### PR TITLE
Add validation for consensus commit mutation operation

### PR DESCRIPTION
**Pull request by Torch3333**
**Context**
There is no validation to ensure Put or Delete operations do not mutate consensus commit metadata columns. Since these columns are only for internal use, users should not be allowed to modify them.

**Changes**
A validation step is added to make sure Put or Delete operations do not mutate consensus commit metadata columns. Additionally, setting operation conditions (see https://github.com/scalar-labs/scalardb/pull/837 ) targeting consensus commit metadata columns are also forbidden.

### CODE REVIEW

File: core/src/main/java/com/scalar/db/transaction/consensuscommit/ConsensusCommitMutationOperationChecker.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    TransactionTableMetadata metadata =
+        transactionTableMetadataManager.getTransactionTableMetadata(operation);
+    if (metadata == null) {
+      throw new IllegalArgumentException(
+          "The specified table is not found: " + operation.forFullTableName().get());
+    }
+    return metadata;
+  }
+
+  public void check(Mutation mutation) throws ExecutionException {
```
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/transaction/consensuscommit/ConsensusCommitMutationOperationChecker.java:39]: I think we can throw `IllegalArgumentException` instead of `ExecutionException` for the operation check.

File: core/src/main/java/com/scalar/db/transaction/consensuscommit/ConsensusCommit.java [hunk 2]
```diff
@@ -79,40 +86,45 @@ public List<Result> scan(Scan scan) throws CrudException {
...
+    checkMutation(put);
     crud.put(put);
   }
 
   @Override
-  public void put(List<Put> puts) {
+  public void put(List<Put> puts) throws CrudException {
     checkArgument(puts.size() != 0);
-    puts.forEach(this::put);
+    for (Put p : puts) {
```
Comment (komamitsu) [inline core/src/main/java/com/scalar/db/transaction/consensuscommit/ConsensusCommit.java:98]: [just out of curiosity] Is this change for suppressing warning or something?

File: core/src/main/java/com/scalar/db/transaction/consensuscommit/ConsensusCommitMutationOperationChecker.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+  }
+
+  private void check(Put put) throws ExecutionException {
+    put.getCondition()
+        .ifPresent(
+            condition -> {
+              if (!(condition instanceof PutIf
+                  || condition instanceof PutIfNotExists
+                  || condition instanceof PutIfExists)) {
+                throw new IllegalArgumentException("The condition is not allowed on Put operation");
```
Comment (komamitsu) [inline core/src/main/java/com/scalar/db/transaction/consensuscommit/ConsensusCommitMutationOperationChecker.java:54]: I guess a user would be interested in the type of the wrong `condition` when receiving this exception. How about showing it in the error message?

File: core/src/main/java/com/scalar/db/transaction/consensuscommit/ConsensusCommitMutationOperationChecker.java (hunk 1 above)
Comment (Torch3333) [reply core/src/main/java/com/scalar/db/transaction/consensuscommit/ConsensusCommitMutationOperationChecker.java:39]: Just to be sure we are on the same page. It is not documented but `check` only throws an `ExceptionException` if the table metadata fails to be retrieved. https://github.com/scalar-labs/scalardb/blob/0212c8b2d8bf773600383687e59f386ce5260faf/core/src/main/java/com/scalar/db/transaction/consensuscommit/ConsensusCommitMutationOperationChecker.java#L30-L31
For validation errors, `IllegalArgumentException` is thrown. 
Do you think we should catch the `ExecutionException` and rethrow it as an `IllegalArgumentException`?

File: core/src/main/java/com/scalar/db/transaction/consensuscommit/ConsensusCommit.java (hunk 2 above)
Comment (Torch3333) [reply core/src/main/java/com/scalar/db/transaction/consensuscommit/ConsensusCommit.java:98]: `this::put` throws the checked exception `CrudException` but the Consumer functional interface does not throw any checked exception.
```
public void forEach(java.util.function.Consumer<? super T> action)
```
Our only option is to use a non-lambda for each if we want to propagate the `CrudException`.

File: core/src/main/java/com/scalar/db/transaction/consensuscommit/ConsensusCommitMutationOperationChecker.java (hunk 3 above)
Comment (Torch3333) [reply core/src/main/java/com/scalar/db/transaction/consensuscommit/ConsensusCommitMutationOperationChecker.java:54]: Yes, that is a good point, thank you. I will add it.

File: core/src/main/java/com/scalar/db/transaction/consensuscommit/ConsensusCommitMutationOperationChecker.java (hunk 1 above)
Comment (brfrn169) [reply core/src/main/java/com/scalar/db/transaction/consensuscommit/ConsensusCommitMutationOperationChecker.java:39]: Ah sorry, I misunderstood. I thought `ExecutionException` was also thrown for validation errors. Please ignore the comment.

File: core/src/main/java/com/scalar/db/transaction/consensuscommit/ConsensusCommitMutationOperationChecker.java (hunk 3 above)
Comment (Torch3333) [reply core/src/main/java/com/scalar/db/transaction/consensuscommit/ConsensusCommitMutationOperationChecker.java]: Revised in https://github.com/scalar-labs/scalardb/pull/873/commits/4e05db8de6de4d94bc81de22ad3c3b40bfd7a321

File: core/src/main/java/com/scalar/db/transaction/consensuscommit/ConsensusCommitMutationOperationChecker.java (hunk 1 above)
Comment (Torch3333) [reply core/src/main/java/com/scalar/db/transaction/consensuscommit/ConsensusCommitMutationOperationChecker.java:46]: I added Javadoc since it's difficult to understand why the `ExecutionException` is thrown.
https://github.com/scalar-labs/scalardb/pull/873/commits/4e05db8de6de4d94bc81de22ad3c3b40bfd7a321

**Review by komamitsu — APPROVED** [review]
LGTM! 👍

File: core/src/main/java/com/scalar/db/transaction/consensuscommit/ConsensusCommitManager.java [hunk 4]
```diff
@@ -180,7 +180,9 @@ DistributedTransaction begin(String txId, Isolation isolation, SerializableStrat
         new Snapshot(txId, isolation, strategy, tableMetadataManager, parallelExecutor);
     CrudHandler crud =
         new CrudHandler(storage, snapshot, tableMetadataManager, isIncludeMetadataEnabled);
-    ConsensusCommit consensus = new ConsensusCommit(crud, commit, recovery);
+    ConsensusCommitMutationOperationChecker operationChecker =
+        new ConsensusCommitMutationOperationChecker(tableMetadataManager);
```
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/transaction/consensuscommit/ConsensusCommitManager.java:184]: Instead of creating a `ConsensusCommitMutationOperationChecker` instance each time, can we hold it as an instance field of `ConsensusCommitManager` and pass it to `ConsensusCommit` instances?

**Review by brfrn169 — COMMENTED** [review]
Left one minor comment. Other than that, LGTM. Please take a look when you have time!

File: core/src/main/java/com/scalar/db/transaction/consensuscommit/ConsensusCommitMutationOperationChecker.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    } else if (mutation instanceof Delete) {
+      check((Delete) mutation);
+    }
+  }
+
+  private void check(Put put) throws ExecutionException {
+    put.getCondition()
+        .ifPresent(
+            condition -> {
+              if (!(condition instanceof PutIf
```
Comment (feeblefakie) [inline core/src/main/java/com/scalar/db/transaction/consensuscommit/ConsensusCommitMutationOperationChecker.java:58]: Just quick confirmation.
This check will be deleted in another PR, which makes conditions usable, right?

**Review by feeblefakie — APPROVED** [review]
LGTM (except for what Toshi pointed out)! Thank you!
I also left a question for just confirmation.

File: core/src/main/java/com/scalar/db/transaction/consensuscommit/ConsensusCommitManager.java (hunk 4 above)
Comment (Torch3333) [reply core/src/main/java/com/scalar/db/transaction/consensuscommit/ConsensusCommitManager.java]: That is a good point, thank you. 
My bad, I did it for the 2pc transaction manager but not for this one. 
Revised in https://github.com/scalar-labs/scalardb/pull/873/commits/de06ddcebf8e54237bba0b7cfbc61935d00c8035

**Review by brfrn169 — APPROVED** [review]
LGTM! Thank you!
